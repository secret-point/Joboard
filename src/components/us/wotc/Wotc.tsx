import React, { useEffect } from "react";
import queryString from "query-string";
import { Col } from "@amzn/stencil-react-components/layout";
import { Text } from "@amzn/stencil-react-components/text";
import IFrame from "../../common/IFrame";
import { useLocation } from "react-router-dom";
import {
  getPageNameFromPath,
  parseQueryParamsArrayToSingleItem,
  resetIsPageMetricsUpdated
} from "../../../helpers/utils";
import { ApplicationState } from "../../../reducers/application.reducer";
import { checkAndBoundGetApplication, getLocale } from "../../../utils/helper";
import { connect } from "react-redux";
import { Locale } from "../../../utils/types/common";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { translate as t } from "../../../utils/translator";
import { JobState } from "../../../reducers/job.reducer";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { ScheduleState } from "../../../reducers/schedule.reducer";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { boundGetScheduleDetail } from "../../../actions/ScheduleActions/boundScheduleActions";
import { boundGetJobDetail } from "../../../actions/JobActions/boundJobDetailActions";

interface MapStateToProps {
  application: ApplicationState;
  job: JobState;
  candidate: CandidateState;
  schedule: ScheduleState;
}

const LocaleToWotcLangMapping: Record<Locale, string> = {
  [Locale.enGB]: "en",
  [Locale.enUS]: "en",
  [Locale.esUS]: "es",
  [Locale.esMX]: "es",
};

export const Wotc = (props: MapStateToProps) => {
  const { application, candidate, schedule, job } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId, jobId, scheduleId } = queryParams;
  const applicationData = application.results;
  const { scheduleDetail } = schedule.results;
  const jobDetail = job.results;
  const candidateData = candidate.results;

  useEffect(() => {
    boundGetCandidateInfo();
  }, []);

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    scheduleId && scheduleId!== scheduleDetail?.scheduleId && boundGetScheduleDetail({
      locale: getLocale(),
      scheduleId: scheduleId
    });
  }, [scheduleId]);

  useEffect(() => {
    jobId && jobId !== jobDetail?.jobId && boundGetJobDetail({ jobId: jobId, locale: getLocale() });
  }, [jobDetail, jobId]);

  useEffect(() => {
    jobDetail && applicationData && candidateData && scheduleDetail && addMetricForPageLoad(pageName);

  }, [jobDetail, applicationData, candidateData, scheduleDetail]);

  useEffect(() => {
    return () => {
      // reset this so as it can emit new pageload event after being unmounted.
      resetIsPageMetricsUpdated(pageName);
    };
  }, []);

  const wotcLocalizedUrl = (wotcUrl: string): string => {
    if (!wotcUrl) {
      return "";
    }
    return `${wotcUrl}&lang=${LocaleToWotcLangMapping[getLocale()]}`;
  };

  return (
    <Col gridGap="S300" padding={{ top: "S300" }}>
      <Text fontSize="T500">
        {t("BB-WOTC-page-title-text", "Work opportunity tax credit questionnaire")}
      </Text>

      <Text>
        {t("BB-WOTC-page-description-text", "Amazon IAT testing participates in federal and or state tax credit programs. The information you give will be used to determine the company's eligibility for these programs and will in no way negatively impact any hiring, retention or promotion decisions.")}
      </Text>

      {applicationData?.wotcScreening?.wotcUrl && (
        <IFrame src={wotcLocalizedUrl(applicationData?.wotcScreening?.wotcUrl)} />
      )}
    </Col>
  );
};

const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(Wotc);
