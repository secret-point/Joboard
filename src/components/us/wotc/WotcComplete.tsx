import React, { useEffect } from "react";
import queryString from "query-string";
import { Col } from "@amzn/stencil-react-components/layout";
import { useLocation } from "react-router";
import { getPageNameFromPath, parseQueryParamsArrayToSingleItem } from "../../../helpers/utils";
import { ApplicationState } from "../../../reducers/application.reducer";
import { boundGetApplication } from "../../../actions/ApplicationActions/boundApplicationActions";
import { boundUpdateWotcStatus } from "../../../actions/WotcActions/boundWotcAction";
import { checkAndBoundGetApplication, getLocale } from "../../../utils/helper";
import { connect } from "react-redux";
import { addMetricForPageLoad } from "../../../actions/AdobeActions/adobeActions";
import { Application, UpdateWotcStatusRequest } from "../../../utils/types/common";
import { CandidateState } from "../../../reducers/candidate.reducer";
import { boundGetCandidateInfo } from "../../../actions/CandidateActions/boundCandidateActions";
import { onCompleteTaskHelper } from "../../../actions/WorkflowActions/workflowActions";

interface MapStateToProps {
  candidate: CandidateState;
  application: ApplicationState;
};

const WOTC_COMPLETE_STATUS = "Completed";

export const WotcComplete = (props: MapStateToProps) => {
  const { candidate, application } = props;
  const { search, pathname } = useLocation();
  const pageName = getPageNameFromPath(pathname);
  const queryParams = parseQueryParamsArrayToSingleItem(queryString.parse(search));
  const { applicationId } = queryParams;
  const applicationData = application.results;
  const { candidateData } = candidate.results;

  useEffect(() => {
    boundGetCandidateInfo();
  }, [])

  useEffect(() => {
    checkAndBoundGetApplication(applicationId);
  }, [applicationId]);

  useEffect(() => {
    applicationData && addMetricForPageLoad(pageName);
  }, [applicationData, pageName]);

  useEffect(() => {
    if (candidateData) {
      const request: UpdateWotcStatusRequest = {
        applicationId,
        candidateId: candidateData?.candidateId,
        status: WOTC_COMPLETE_STATUS
      };

      // TODO: remove any
      boundUpdateWotcStatus(request, (data: any) => {
        // get the latest application data and update the application state, then complete the step
        boundGetApplication({ applicationId: applicationId, locale: getLocale() }, (app: Application) => {
          onCompleteTaskHelper(app);
        });
      });
    }

  }, [applicationId, candidateData]);


  return (
    <Col minHeight="40vh"></Col>
  )
}

const mapStateToProps = (state: MapStateToProps) => {
  return state;
};

export default connect(mapStateToProps)(WotcComplete);
