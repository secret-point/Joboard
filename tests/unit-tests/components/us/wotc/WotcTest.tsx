import React from "react";
import { shallow } from "enzyme";
import routeData from 'react-router';
import { Wotc } from "../../../../../src/components/us/wotc/Wotc";
import { TEST_APPLICATION_ID, TEST_APPLICATION_STATE, TEST_CANDIDATE_STATE, TEST_JOB_ID, TEST_JOB_STATE, TEST_SCHEDULE_ID, TEST_SCHEDULE_STATE } from "../../../../test-utils/test-data";

describe("Wotc", () => {
  const mockLocation = {
    pathname: "/wotc",
    search: `?jobId=${TEST_JOB_ID}&applicationId=${TEST_APPLICATION_ID}&scheduleId=${TEST_SCHEDULE_ID}`,
    hash: '',
    state: null
  };

  it("should match snapshot", () => {
    jest.spyOn(routeData, 'useLocation').mockReturnValue(mockLocation);

    const shallowWrapper = shallow(
      <Wotc
        candidate={TEST_CANDIDATE_STATE}
        job={TEST_JOB_STATE}
        application={TEST_APPLICATION_STATE}
        schedule={TEST_SCHEDULE_STATE}
      />);

    expect(shallowWrapper).toMatchSnapshot();
  });
});
