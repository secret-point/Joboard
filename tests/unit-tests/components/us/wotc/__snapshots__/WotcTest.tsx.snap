// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wotc should match snapshot 1`] = `
<Col
  gridGap="S300"
  padding={
    Object {
      "top": "S300",
    }
  }
>
  <Text
    font="primary"
    fontSize="T500"
    hyphens="none"
  >
    Work opportunity tax credit questionnaire
  </Text>
  <Text
    font="primary"
    fontSize="T300"
    hyphens="none"
  >
    Amazon IAT testing participates in federal and or state tax credit programs. The information you give will be used to determine the company's eligibility for these programs and will in no way negatively impact any hiring, retention or promotion decisions.
  </Text>
</Col>
`;
